torchmdnet/models/tensornet.py

```diff
@@ -188,19 +205,28 @@ def forward(
         q: Optional[Tensor] = None,
         s: Optional[Tensor] = None,
     ) -> Tuple[Tensor, Optional[Tensor], Tensor, Tensor, Tensor]:
-
         # Obtain graph, with distances and relative position vectors
         edge_index, edge_weight, edge_vec = self.distance(pos, batch)
         # This assert convinces TorchScript that edge_vec is a Tensor and not an Optional[Tensor]
         assert (
             edge_vec is not None
         ), "Distance module did not return directional information"
-
-        # Expand distances with radial basis functions
+        # Distance module returns -1 for non-existing edges, to avoid having to resize the tensors when we want to ensure static shapes (for CUDA graphs) we make all non-existing edges pertain to the first atom
+        if self.static_shapes:
+            mask = (edge_index[0] >= 0).unsqueeze(0).expand_as(edge_index)
+            # I trick the model into thinking that the masked edges pertain to the first atom
+            # WARNING: This can hurt performance if max_num_pairs >> actual_num_pairs
+            edge_index = edge_index * mask
+            edge_weight = edge_weight * mask[0]
+            edge_vec = edge_vec * mask[0].unsqueeze(-1).expand_as(edge_vec)
         edge_attr = self.distance_expansion(edge_weight)
-        # Embedding from edge-wise tensors to node-wise tensors
+        mask = edge_index[0] != edge_index[1]
+        # Normalizing edge vectors by their length can result in NaNs, breaking Autograd.
+        # I avoid dividing by zero by setting the weight of self edges and self loops to 1
+        edge_vec = edge_vec / torch.masked_scatter(torch.ones_like(edge_weight),
```

Comment:
I think `maked_scatter` is working differently than you think (https://pytorch.org/docs/stable/generated/torch.Tensor.masked_scatter.html)

This should be a correct implementation
```
diff --git a/torchmdnet/models/tensornet.py b/torchmdnet/models/tensornet.py
index 045219b..6b005b1 100644
--- a/torchmdnet/models/tensornet.py
+++ b/torchmdnet/models/tensornet.py
@@ -220,12 +220,10 @@ class TensorNet(nn.Module):
             edge_weight = edge_weight * mask[0]
             edge_vec = edge_vec * mask[0].unsqueeze(-1).expand_as(edge_vec)
         edge_attr = self.distance_expansion(edge_weight)
-        mask = edge_index[0] != edge_index[1]
         # Normalizing edge vectors by their length can result in NaNs, breaking Autograd.
         # I avoid dividing by zero by setting the weight of self edges and self loops to 1
-        edge_vec = edge_vec / torch.masked_scatter(torch.ones_like(edge_weight),
-            mask, edge_weight
-        ).unsqueeze(1)
+        mask = edge_index[0] == edge_index[1]
+        edge_vec = edge_vec / edge_weight.masked_fill(mask, 1).unsqueeze(1)
         X = self.tensor_embedding(z, edge_index, edge_weight, edge_vec, edge_attr)
         for layer in self.layers:
             X = layer(X, edge_index, edge_weight, edge_attr)

```

Reply:
The documentation is actually super misleading I believe https://github.com/pytorch/pytorch/issues/82377

Reply:
But looking at it your implementation looks better and masked_scatter is problematic with torch.compile. Lets use yours. Thanks!